File: exercise11-12/utilities_for_exercise11-12/MATLAB/load_my_digit.m
function [my_digit] = load_my_digit(fpath)
my_digits = imread(fpath);
% Image (containing many digits wrote on white papers) pre-processing
my_digits = rgb2gray(my_digits); % converting to gray-scale
g=figure('Units','normalized','Position',[0 0 .5 .5]);
imagesc(my_digits); colormap gray
title('Select two points: top-left and bottom-right corner of the area containing a digit of interest...')
% Select the top-left and bottom-right corner of the area containing a digit
[x,y] = ginput(2);
close(g)

my_digits = 255-my_digits; % inverting dynamic (higher values: foreground, low values: background)
my_digits(my_digits<128)=0; % roughly bringing to 0 background
my_digits = double(my_digits)/255; % normalization between 0-1

% Digit pre-processing
start_row = int16(y(1));
stop_row = int16(y(2));
start_col = int16(x(1));
stop_col = int16(x(2));
% cropping the selected digit
my_digit = my_digits(start_row:stop_row, start_col:stop_col);
% Resizing digits according to the adopted dataset
my_digit = imresize(my_digit, [28, 28]);
my_digit(my_digit<0)=0;
my_digit(my_digit>1)=1;
% Forcing one channel (gray-scale image)
my_digit = reshape(my_digit, [28, 28, 1]);
end

File: exercise11-12/utilities_for_exercise11-12/Python/load_my_digit.py
import numpy as np
from PIL import Image


def rgb2gray(rgb):
    ''' Function that returns a gray-scale image from an input RGB image

     Arguments
     ---------
     rgb: ndarray
           Numpy array containing the RGB image (3-D).
    '''

    r, g, b = rgb[:,:,0], rgb[:,:,1], rgb[:,:,2]
    gray = 0.2989 * r + 0.5870 * g + 0.1140 * b
    return gray


def load_my_digit(fpath):
    ''' Load and pre-process a handwritten digit contained in a
     picture. The digit is returned as a 28x28 uint8 ndarray.

     Arguments
     ---------
     fpath: string
           Filepath of the target image.

     Author
     ------
     Davide Borra, 2022
    '''
    my_digit = Image.open(fpath)
    my_digit = np.array(my_digit)
    my_digit = rgb2gray(my_digit)

    my_digit = 255-my_digit # inverting dynamic (higher values: foreground, low values: background)
    my_digit[my_digit<128]=0 # roughly bringing to 0 all background

    my_digit = np.array(Image.fromarray(my_digit).resize((28,28)))
    my_digit[my_digit<0] = 0 # saturating to 0 and 255 interpolated values exceeding 0-255 range
    my_digit[my_digit>255] = 255
    return my_digit.astype(np.uint8)
